biscuit-auth/src/error.rs

```diff
@@ -168,9 +171,9 @@ pub enum Signature {
 #[derive(Error, Clone, Debug, PartialEq, Eq)]
 #[cfg_attr(feature = "serde-error", derive(serde::Serialize, serde::Deserialize))]
 pub enum Logic {
-    #[error("a rule provided by a block is generating facts with the authority or ambient tag, or has head variables not used in its body")]
+    #[error("a rule provided by a block is producing a fact with unbound variables")]
```

Comment:
i wanted to completely remove it, but it's still used. the strange thing is that it explicitly mentions block rules. Where is the error for authorizer rules with unbound variables in facts?

Reply:
after a bit of digging:

- unbound variables are checked by the parser when constructing rules, so all rules constructed by the parser will be checked, regardless of their origin, and the error will be a parser error, not a `Logic` error
- this error can indeed only happen when a block is added to the authorizer, so adding a broken rule to the authorizer will not be caught here (rules constructed by the parser will be ok, it can only happen for manually constructed rules)
- rules with unbound variables in head will not generate any fact (but not crash)

I think the ergonomics of this could be improved:

- by being clearer on invariants that `Rule` and friends should guarantee, no matter how they were constructed (this would be a bit of work because i think that we should do it in other places as well)
- for now, by calling `validate_variables` when adding rules to the authorizer, not only rules blocks

Reply:
That makes sense, we should avoid those kinds of footguns